Comparing our methods to Jorrin and Imperial 2015 paper:

Methods we have done so far: Filter reads with Trimmomatic (Kbase), align reads to Rlv reference genome, transform output with Sam Tools, detect SNPs (They used varscan, we used PoPoolation)

Methods they did, which we should do next: Filter SNPs bassed on requirements (FST > 0.1, p < 0.05), calculate euclidean distances and multidimensional scaling based on output
